File: web/src/models/Order.ts
import mongoose, { Document, Schema } from 'mongoose';

export interface IOrderItem {
  product: mongoose.Types.ObjectId;
  qty: number;
  color?: string;
  lensType?: string;
  lensSubType?: string;
  power?: {
    RE?: { sph?: number; cyl?: number; axis?: number };
    LE?: { sph?: number; cyl?: number; axis?: number };
    pd?: number;
  };
  lensQuality?: string;
  lensPrice?: number;
  framePrice: number;
  fittingCharge: number;
}

export interface IOrder extends Document {
  orderNumber: string;
  orderId: string;
  user: mongoose.Types.ObjectId;
  items: IOrderItem[];
  address: {
    fullName: string;
    mobile: string;
    line1: string;
    line2?: string;
    city: string;
    state: string;
    pincode: string;
  };
  subtotal: number;
  deliveryCharge: number;
  fittingCharge: number;
  discount: number;
  total: number;
  coupon?: {
    code: string;
    discountType: 'percent' | 'flat';
    discountValue: number;
    amountSaved: number;
  };
  paymentStatus: 'pending' | 'paid' | 'failed' | 'refunded';
  paymentMethod?: string;
  transactionId?: string;
  status: 'pending' | 'confirmed' | 'processing' | 'shipped' | 'delivered' | 'cancelled' | 'returned';
  statusHistory: { status: string; timestamp: Date; note?: string }[];
  trackingNumber?: string;
  courierPartner?: string;
  estimatedDelivery?: Date;
  prescriptionVerified: boolean;
  internalNotes: { note: string; addedBy?: mongoose.Types.ObjectId; addedAt: Date }[];
  isFlagged: boolean;
  createdAt: Date;
  updatedAt: Date;
}

const PowerSchema = new Schema(
  {
    RE: { sph: Number, cyl: Number, axis: Number },
    LE: { sph: Number, cyl: Number, axis: Number },
    pd: Number,
  },
  { _id: false }
);

const OrderItemSchema = new Schema<IOrderItem>({
  product: { type: Schema.Types.ObjectId, ref: 'Product' },
  qty: { type: Number, default: 1 },
  color: String,
  lensType: String,
  lensSubType: String,
  power: PowerSchema,
  lensQuality: String,
  lensPrice: Number,
  framePrice: { type: Number, default: 1 },
  fittingCharge: { type: Number, default: 0 },
});

const OrderSchema = new Schema<IOrder>(
  {
    orderNumber: { type: String, unique: true },
    orderId: { type: String, unique: true },
    user: { type: Schema.Types.ObjectId, ref: 'User', required: true },
    items: [OrderItemSchema],
    address: {
      fullName: String,
      mobile: String,
      line1: String,
      line2: String,
      city: String,
      state: String,
      pincode: String,
    },
    subtotal: Number,
    deliveryCharge: { type: Number, default: 99 },
    fittingCharge: { type: Number, default: 0 },
    discount: { type: Number, default: 0 },
    total: Number,
    coupon: {
      code: String,
      discountType: { type: String, enum: ['percent', 'flat'] },
      discountValue: Number,
      amountSaved: Number,
    },
    paymentStatus: {
      type: String,
      enum: ['pending', 'paid', 'failed', 'refunded'],
      default: 'pending',
    },
    paymentMethod: String,
    transactionId: String,
    status: {
      type: String,
      enum: ['pending', 'confirmed', 'processing', 'shipped', 'delivered', 'cancelled', 'returned'],
      default: 'pending',
    },
    statusHistory: [
      {
        status: String,
        timestamp: { type: Date, default: Date.now },
        note: String,
      },
    ],
    trackingNumber: String,
    courierPartner: String,
    estimatedDelivery: Date,
    prescriptionVerified: { type: Boolean, default: false },
    internalNotes: [
      {
        note: String,
        addedBy: { type: Schema.Types.ObjectId, ref: 'User' },
        addedAt: { type: Date, default: Date.now },
      },
    ],
    isFlagged: { type: Boolean, default: false },
  },
  { timestamps: true }
);

export const Order = mongoose.models.Order || mongoose.model<IOrder>('Order', OrderSchema);

File: mobile/lib/models/order.dart
class OrderItem {
  final String? productName;
  final String? color;
  final String? lensType;
  final String? lensQuality;
  final double framePrice;
  final double? lensPrice;
  final double? fittingCharge;
  final int qty;
  final String? productImage;
  final String? sku;

  OrderItem({
    this.productName,
    this.color,
    this.lensType,
    this.lensQuality,
    this.framePrice = 1,
    this.lensPrice,
    this.fittingCharge,
    this.qty = 1,
    this.productImage,
    this.sku,
  });

  factory OrderItem.fromJson(Map<String, dynamic> json) => OrderItem(
        productName: json['product'] is Map ? json['product']['name'] : null,
        color: json['color'],
        lensType: json['lensType'],
        lensQuality: json['lensQuality'],
        framePrice: (json['framePrice'] as num?)?.toDouble() ?? 1,
        lensPrice: (json['lensPrice'] as num?)?.toDouble(),
        fittingCharge: (json['fittingCharge'] as num?)?.toDouble(),
        qty: json['qty'] ?? 1,
        productImage: json['product'] is Map && json['product']['images'] is List && (json['product']['images'] as List).isNotEmpty ? (json['product']['images'] as List).first.toString() : null,
        sku: json['product'] is Map ? json['product']['sku']?.toString() : null,
      );
}

class OrderAddress {
  final String fullName;
  final String mobile;
  final String line1;
  final String? line2;
  final String city;
  final String state;
  final String pincode;

  OrderAddress({
    required this.fullName,
    required this.mobile,
    required this.line1,
    this.line2,
    required this.city,
    required this.state,
    required this.pincode,
  });

  factory OrderAddress.fromJson(Map<String, dynamic> json) => OrderAddress(
        fullName: json['fullName']?.toString() ?? '',
        mobile: json['mobile']?.toString() ?? '',
        line1: json['line1']?.toString() ?? '',
        line2: json['line2']?.toString(),
        city: json['city']?.toString() ?? '',
        state: json['state']?.toString() ?? '',
        pincode: json['pincode']?.toString() ?? '',
      );
}

class Order {
  final String id;
  final String orderNumber;
  final List<OrderItem> items;
  final String status;
  final double total;
  final double subtotal;
  final double deliveryCharge;
  final double fittingCharge;
  final double discount;
  final double walletUsed;
  final String createdAt;
  final OrderAddress? address;
  final String paymentStatus;
  final String? paymentMethod;
  final String? trackingNumber;
  final String? courierPartner;
  final String? estimatedDelivery;

  Order({
    required this.id,
    required this.orderNumber,
    this.items = const [],
    this.status = 'pending',
    this.total = 0,
    this.subtotal = 0,
    this.deliveryCharge = 99,
    this.fittingCharge = 0,
    this.discount = 0,
    this.walletUsed = 0,
    this.createdAt = '',
    this.address,
    this.paymentStatus = 'pending',
    this.paymentMethod,
    this.trackingNumber,
    this.courierPartner,
    this.estimatedDelivery,
  });

  factory Order.fromJson(Map<String, dynamic> json) => Order(
        id: json['_id'] ?? json['id'] ?? '',
        orderNumber: json['orderId'] ?? '',
        items: (json['items'] as List<dynamic>?)
                ?.map((i) => OrderItem.fromJson(i))
                .toList() ??
            [],
        status: json['status'] ?? 'pending',
        total: (json['total'] as num?)?.toDouble() ?? 0,
        subtotal: (json['subtotal'] as num?)?.toDouble() ?? 0,
        deliveryCharge: (json['deliveryCharge'] as num?)?.toDouble() ?? 99,
        fittingCharge: (json['fittingCharge'] as num?)?.toDouble() ?? 0,
        discount: (json['discount'] as num?)?.toDouble() ?? 0,
        walletUsed: (json['walletUsed'] as num?)?.toDouble() ?? 0,
        createdAt: json['createdAt'] ?? '',
        address: json['address'] != null ? OrderAddress.fromJson(json['address']) : null,
        paymentStatus: json['paymentStatus'] ?? 'pending',
        paymentMethod: json['paymentMethod']?.toString(),
        trackingNumber: json['trackingNumber']?.toString(),
        courierPartner: json['courierPartner']?.toString(),
        estimatedDelivery: json['estimatedDelivery']?.toString(),
      );
}
